common/config/persistence.go

```diff
@@ -97,6 +99,43 @@ func (c *Persistence) Validate() error {
 		if ds.SQL != nil && ds.NoSQL != nil {
 			return fmt.Errorf("persistence config: datastore %v: only one of SQL or NoSQL can be specified", st)
 		}
+		if ds.SQL != nil {
+			if ds.SQL.UseMultipleDatabases {
+				if !useAdvancedVisibilityOnly {
+					return fmt.Errorf("sql persistence config: multipleSQLDatabases can only be used with advanced visibility only")
+				}
+				if ds.SQL.DatabaseName != "" {
+					return fmt.Errorf("sql persistence config: databaseName can only be configured in multipleDatabasesConfig when UseMultipleDatabases is true ")
+				}
+				if ds.SQL.ConnectAddr != "" {
+					return fmt.Errorf("sql persistence config: connectAddr can only be configured ini multipleDatabasesConfig when UseMultipleDatabases is true ")
+				}
+				if ds.SQL.User != "" {
+					return fmt.Errorf("sql persistence config: user can only be configured ini multipleDatabasesConfig when UseMultipleDatabases is true ")
+				}
+				if ds.SQL.Password != "" {
+					return fmt.Errorf("sql persistence config: password can only be configured ini multipleDatabasesConfig when UseMultipleDatabases is true ")
+				}
+				if ds.SQL.NumShards <= 1 || len(ds.SQL.MultipleDatabasesConfig) != ds.SQL.NumShards {
+					return fmt.Errorf("sql persistence config: nShards must be greater than one and equal to the length of multipleDatabasesConfig")
```

Comment:
nit: I wonder if we need `NumShards` or can just rely on length of `MultipleDatabasesConfig` instead?

Reply:
`NumShards` is needed in the case that Cadence uses a sharded SQL solution(like Uber's docStore, or CockroachDB/DorisDB), so we can't remove it(in that case, they don't have `MultipleDatabasesConfig`.)

I was thinking to give  `NumShards` a default number of length of `MultipleDatabasesConfig` when not set. But the doc is going to be a little confusing -- when `MultipleDatabasesConfig` set the default value is length of it, when not set the default value is 1. 

On the other side, `NumShards` is a critical config that having it explicitly will help understanding it when reviewing the config(rather than counting the length the of array). 

So I am not sure whether we should complicate that. But open to more ideas.

Reply:
I merged it now as it's a nit comment. 
But feel free to give suggestion as I will open next PR to tests and keep improving it.